### ISSUE TITLE
vc dev issues with node 17; unresponsive site and/or 502 on api routes (ipv4 / ipv6 issue)

### ISSUE THREAD

**Issue by gadicc** [zeit/now#7306]
**TL;DR;**

The issue is caused by changes in DNS resolution order in node 17.  The quickest short-term *workaround* is to remove `::1` from your `hosts` file (or otherwise change your system to resolve `localhost` to 127.0.0.1 before ::1), but watch this issue for a better solution.

**Full Explanation**

This issue affects `create-react-app` and presumably any other framework which defaults to `ipv4` and where `vc dev` waits for a port to become reachable.

Node 17 has a potentially breaking change where [`--dns-result-order` now defaults to `verbatim`](https://github.com/nodejs/node/commit/1b2749ecbe) (instead of `ip4first`).  Depending on system setup, this means `localhost` will now resolve to `::1` (ipv6) instead of `127.0.0.1` (ipv4).

Since create-react-app [currently defaults to ipv4](https://github.com/facebook/create-react-app/issues/11302), what happens in this case is:

1. `create-react-app` still opens your browser to `localhost` (127.0.0.1) on some random port
2. Browser requests to e.g. `localhost:3000` hang (with little hint why on the console)
3. After 5m (i.e. 300s), `vc dev` fails with, e.g.

```
Error! Detecting port 33409 timed out after 300000ms
```

This is because `vc dev`, using `is-port-reachable`, is testing on `localhost` which in this case resolves to `::1`, while `create-react-app` is only listening on `127.0.0.1`, so it will never be reachable (for `vc dev`).

The easiest workaround is to modify your system to resolve the `ipv4` address first, or, failing that, to remove `::1` from your `hosts` file.  Obviously this is not a good long-term solution.

You can force `create-react-app` to bind to an ipv6 with the `HOST` environment variable:

```
$ HOST=::1 vc dev
```

This allows the dev server to fully load and proxy requests to `create-react-app`, but API routes fail since the serverless funcs don't listen for them on `::1`, i.e.

```
Error! Failed to complete request to /api/serverless: Error: connect ECONNREFUSED ::1:35855
```

That's perhaps a different issue though :)

If relevant, I'm currently on Vercel CLI 23.1.3-canary.72.

A little surprised I'm the first one to report this, but maybe my system is unique :)

**Comment by gadicc** [comment]
In a **Next.js project**, the proxying part works fine but then we get the second error above again for serverless functions:

```
Error! Failed to complete request to /api/serverless: Error: connect ECONNREFUSED ::1:35855
```

Incidentally, on the browser side, the response is:

```
Failed to load resource: the server responded with a status of 502 (Bad Gateway)
```

If you try access the API route directly with a browser, you get the full Vercel error page:

![image](https://user-images.githubusercontent.com/381978/150630220-961b8f21-a07f-4fe8-b19d-59947441ea2b.png)

**502**: BAD_GATEWAY
Code: \`NO_RESPONSE_FROM_FUNCTION\`

**Comment by jbbelcher53** [comment]
@gadicc https://github.com/vercel/og-image/issues/170#issuecomment-1028319698 this worked for me. I was running into the same issue should work if you upgrade to canary "npm i -g vercel@canary"

**Comment by gadicc** [comment]
Hey @jbbelcher53, thanks for this.  Unfortunately I'm still having the same issue even with `23.1.3-canary.73`.

**Comment by piducancore** [comment]
I'm still having this issue with `24.0.1-canary.1` and node `v17.5.0`. The workaround mentioned above (removing `::1` from `/etc/hosts`) resolves it.

**Comment by romke-armor-fi** [comment]
Same issue and solution here on macos, thanks @gadicc !

**Comment by lerayj** [comment]
Same issue too with Nuxt, same solution. This issue lost me 3 days of work, thanks @gadicc for the solution!

**Comment by ZeldOcarina** [comment]
Same issue for me, I'm not sure why tweaking the hosts file and removing localhost helps but it does!
